java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java

```diff
@@ -0,0 +1,627 @@
+    * if setValueCount() has been done else it is always 0. This again implies
+    * that it is probably no effect on general purpose vector operations.
+    */
+
+   protected abstract org.slf4j.Logger getLogger();
+
+   @Override
+   public Mutator getMutator() {
+      throw new  UnsupportedOperationException("Mutator is not needed to write into vector");
+   }
+
+   @Override
+   public Accessor getAccessor() {
+      throw new UnsupportedOperationException("Accessor is not needed to read from vector");
+   }
+
+   @Override
+   public long getValidityBufferAddress() {
+      return (validityBuffer.memoryAddress());
+   }
+
+   @Override
+   public long getDataBufferAddress() {
+      return (valueBuffer.memoryAddress());
+   }
+
+   @Override
+   public long getOffsetBufferAddress() {
+      throw new UnsupportedOperationException("not supported for fixed-width vectors");
+   }
+
+   @Override
+   public ArrowBuf getValidityBuffer() {
+      return validityBuffer;
+   }
+
+   @Override
+   public ArrowBuf getDataBuffer() {
+      return valueBuffer;
+   }
+
+   @Override
+   public ArrowBuf getOffsetBuffer() {
+      throw new UnsupportedOperationException("not supported for fixed-width vectors");
+   }
+
+   @Override
+   public void setInitialCapacity(int valueCount) {
+      final long size = (long)valueCount * typeWidth;
+      if (size > MAX_ALLOCATION_SIZE) {
+         throw new OversizedAllocationException("Requested amount of memory is more than max allowed");
+      }
+      valueAllocationSizeInBytes = (int)size;
+      validityAllocationSizeInBytes = getSizeFromCount(valueCount);
+   }
+
+   @Override
+   public int getValueCapacity(){
+      return Math.min(getValueBufferValueCapacity(), getValidityBufferValueCapacity());
+   }
+
+   /* for test purposes */
+   private int getValueBufferValueCapacity() {
+      return (int)((valueBuffer.capacity() * 1.0)/typeWidth);
+   }
+
+   /* for test purposes */
+   private int getValidityBufferValueCapacity() {
+      return (int)(validityBuffer.capacity() * 8L);
+   }
+
+   /* number of bytes for the validity buffer for the given valueCount */
+   protected int getSizeFromCount(int valueCount) {
+      return (int) Math.ceil(valueCount / 8.0);
+   }
+
+   @Override
+   public void zeroVector() {
+      initValidityBuffer();
+      initValueBuffer();
+   }
+
+   private void initValidityBuffer() {
+      validityBuffer.setZero(0, validityBuffer.capacity());
+   }
+
+   private void initValueBuffer() {
+      valueBuffer.setZero(0, valueBuffer.capacity());
+   }
+
+   public void reset() {
+      zeroVector();
+   }
+
+   @Override
+   public void close() { clear(); }
+
+   @Override
+   public void clear() {
+      releaseBuffer(validityBuffer);
+      releaseBuffer(valueBuffer);
+   }
+
+   /* used to step down the memory allocation */
+   protected void incrementAllocationMonitor() {
+      if (allocationMonitor < 0) {
```

Comment:
Maybe refactor `allocationMonitor` to it's own class?

Reply:
I am trying to err on the side of less objects and less inner classes. So that is why thought we can maintain the allocationMonitor state through a simple counter. What do you think?

Reply:
I think it's fine to not have a object, but maybe put the allocationMonitor logic in base class?

Reply:
Right now we have other vector types inheriting from BaseValueVector. Once the entire hierarchy under BaseValueVector is refactored, we can move few things to BaseValueVector. There are a few things which are duplicated between BaseNullableFixedWidthVector and BaseNullableVariableWidthVector. Later on they can be moved to top level base class  -- BaseValueVector.

Reply:
In that case, I am leaning towards refactor allocationMonitor logic to a standalone class (in the logic is indeed the same among different vectors). I'd rather not maintaining the same logic in many different places...

Reply:
Let me clarify. I am aware that there is handful logic which is duplicated between two abstract classes only -- BaseNullableFixedWidthVector and BaseNullableVariableWidthVector. This is intentional for the time being.

It is not spread across several classes. We can move this shared logic to top level class BaseValueVector once the entire hierarchy has been refactored. 

For example, we haven't touched classes like ListVector and other which also inherit from BaseValueVector.

Reply:
I am just trying to phase out the work. Does the above explanation make some sense?

Reply:
Yes it does make sense. It's fine. We can always refactor in a later patch if we want to.

Reply:
That sounds good to me @siddharthteotia to move it to a top level class later.  Do you think you could add a comment in the places that you anticipate changing in later phases?

Reply:
Sure, I will add note(s) in the code. Thanks @BryanCutler  and @icexelloss

Reply:
Done.